Python/symtable.c

```diff
@@ -82,6 +82,7 @@ ste_new(struct symtable *st, identifier name, _Py_block_ty block,
 
     if (st->st_cur != NULL &&
         (st->st_cur->ste_nested ||
+         // TODO and AnnotationBlock?
```

Comment:
First, I assume you meant "or AnnotationBlock?" here.  So you're wondering if it the code should be amended to:

    if (st->st_cur != NULL &&
        (st->st_cur->ste_nested ||
         st->st_cur->ste_type == FunctionBlock ||
         st->st_cur->ste_type == AnnotationBlock))
        ste->ste_nested = 1;

The answer is: nope, no point.  This is from `ste_new()`, where we're creating a new block that's gonna be below (enclosed by) `st->st_cur`.  But `AnnotationBlock` never enclose other blocks.  So `st->st_cur->ste_type` could never be `AnnotationBlock` QED.

(Note that I just spent 45 minutes figuring this out, and I was pretty convinced the answer was "yes" for a while.)

Reply:
Thanks for staring at this!

But an AnnotationBlock *could* enclose another block, if you put a lambda in your annotation. Not that I'd recommend it, but we should make sure it works correctly for some definition of "correctly".

Reply:
That's an interesting point.  I hadn't thought of lambdas there.  But there's another reason I think the answer is "it doesn't matter".  The only place `ste_nested` is used for anything is at the end of `analyze_name()`.  After failing to figure out anything else about the name in question: if the current block is `ste_nested`, it marks the block as `ste_free`, which means "this block has any free variables".

Long story short: on the one hand, it seems reasonable to assume this name is a "free variable".  But on the other, more important hand, `ste_free` is never used for anything.  Well, I mean, it's used to set the `ste_child_free` flag on all enclosing scopes.  But `ste_child_free` is never used for anything.  Which means `ste_free` is irrelevant.  Which means `ste_nested` is also irrelevant.

`ste_nested` *is* visible from Python; it's propogated to the code object (see `compile.c` function `compute_code_flags()`).  So maybe AnnotationBlock code objects should also have it set?  Maybe `compute_code_flags()` needs to change to

    if ((ste->ste_type == FunctionBlock) || (ste->ste_type == AnnotationBlock))

In the final analysis, it seems like in the spirit of the code, it should be a nested block.  Even if it doesn't actually matter.

(Am I wrong?  Did I miss a way in which either `ste_nested`, `ste_free`, or `ste_child_free` is ever used for anything important?  I can't see any other hits by grepping through the whole Python tree.  Symbol table objects are visible from inside Python, but those specific fields can't be queried from Python on an ste.)

Below here I'll leave in the "long story", the analysis I wrote on the topic before I noticed it wasn't used for anything.

--

`ste_nested` only actually has an effect in one spot.  It's at the end of `analyze_name()`.  The code is trying to figure out what kind of variable `name` is.  At this point we've already established

* the name hasn't been explicitly declared global or non-local in the current block/scope,
* the name hasn't been bound in the current block/scope,
* nor have we seen a binding for the name in a visible enclosing scope, and
* the name hasn't been explicitly declared global 

I'm not sure where the name can come from at that point.  Maybe it's a builtin.  Maybe it's from "from foo import *", which means it's a global (because "from foo import *" is only legal in module scope in Python 3).

Anyway, this is a last-ditch analysis about this name we just don't know anything about.  And so the code says:

    if (ste->ste_nested)
        ste->ste_free = 1;
    SET_SCOPE(scopes, name, GLOBAL_IMPLICIT);

If we're in a nested scope, then also mark the scope as having free variables.  Globals  it's certainly a *free variable.*